### ISSUE TITLE
CI: Failing RestIT on master (nodes.stats/30_discovery/Discovery stats) 

### ISSUE THREAD

**Issue by cbuescher** [elastic/elasticsearch#18433]
This Rest test started failing a few hours ago, I have seen it fail several times on master branch:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/415/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/844/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-intake/583/console

```
:distribution:integ-test-zip:integTest
Executing task ':distribution:integ-test-zip:integTest' (up-to-date check took 0.001 secs) due to:
  Task.upToDateWhen is false.
[ant:junit4] <JUnit4> says kaixo! Master seed: D34B10548736FEA
==> Test Info: seed=D34B10548736FEA; jvm=1; suite=1
Started J0 PID(342@slave-886fdd4f.build.us-west-2a.elasticnet.co).
Suite: org.elasticsearch.test.rest.RestIT
IGNOR/A 0.01s | RestIT.test {p0=update/85_fields_meta/Metadata Fields}
   > Assumption #1: [update/85_fields_meta/Metadata Fields] skipped, reason: [Update doesn't return metadata fields, waiting for #3259] 
  2> REPRODUCE WITH: gradle :distribution:integ-test-zip:integTest -Dtests.seed=D34B10548736FEA -Dtests.class=org.elasticsearch.test.rest.RestIT -Dtests.method="test {p0=nodes.stats/30_discovery/Discovery stats}" -Des.logger.level=WARN -Dtests.security.manager=true -Dtests.jvm.argline="-XX:-UseParNewGC -XX:-UseConcMarkSweepGC -XX:+UseG1GC" -Dtests.locale=el-CY -Dtests.timezone=America/Chihuahua
FAILURE 0.01s | RestIT.test {p0=nodes.stats/30_discovery/Discovery stats} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: field [nodes] doesn't have a true value
   > Expected: not null
   >      but: was null
   > 	at __randomizedtesting.SeedInfo.seed([D34B10548736FEA:85608EDFE68F0212]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.test.rest.section.IsTrueAssertion.doAssert(IsTrueAssertion.java:48)
   > 	at org.elasticsearch.test.rest.section.Assertion.execute(Assertion.java:69)
   > 	at org.elasticsearch.test.rest.ESRestTestCase.test(ESRestTestCase.java:399)
   > 	at java.lang.Thread.run(Thread.java:745)
  1> [2016-05-18 04:56:46,232][ERROR][org.elasticsearch.test.rest.client] Adding header Accept
  1>  with value application/yaml
IGNOR/A 0.00s | RestIT.test {p0=indices.put_settings/all_path_options/put settings in list of indices}
   > Assumption #1: [indices.put_settings/all_path_options/put settings in list of indices] skipped, reason: [list of indices not implemented yet] 
  1> [2016-05-18 04:57:21,422][ERROR][org.elasticsearch.test.rest.client] Adding header Accept
  1>  with value application/yaml
  2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+master+g1gc/distribution/integ-test-zip/build/testrun/integTest/J0/temp/org.elasticsearch.test.rest.RestIT_D34B10548736FEA-001
  2> NOTE: test params are: codec=Asserting(Lucene60), sim=RandomSimilarity(queryNorm=true,coord=no): {}, locale=el-CY, timezone=America/Chihuahua
  2> NOTE: Linux 2.6.32-573.el6.x86_64 amd64/Oracle Corporation 1.8.0_71 (64-bit)/cpus=4,threads=1,free=275366168,total=536870912
  2> NOTE: All tests run in this JVM: [RestIT]
Completed [1/1] in 73.34s, 519 tests, 1 failure, 2 skipped <<< FAILURES!

Tests with failures:
  - org.elasticsearch.test.rest.RestIT.test {p0=nodes.stats/30_discovery/Discovery stats}
```

Also reproduces locally when I run `gradle clean check` in `distribution/integ-test-zip`, but cannot reproduce with the above "REPRODUCE WITH" line.

**Comment by cbuescher** [comment]
I temporarily disabled the test with de321fb159f4e9f764d3fe6f1c03c664bf15b384.

**Comment by clintongormley** [comment]
@bleskes @ywelsch this is to do with node/stats/discovery - something you've touched recently?

**Comment by bleskes** [comment]
@tlrx thx for fixing. I wonder if we need to add something to our rest failure messages/reporting as I know  @cbuescher spent some time trying to figure things out. Outputting/incorrect/unbalanced JSON should have been obvious...  How did you end up finding it?

**Comment by tlrx** [comment]
@bleskes I reproduced the test in DEBUG mode so that it prints the JSON response, which showed that the `nodes` fields was missing (as stated in the test failure message). 

The thing to know with `filter_path` expressions like `nodes.*.discovery` is that the parent fields corresponding to `nodes` and `*` are written in the response only if at least one child field matches. So `nodes` is written only if `nodes.ABC.discovery` exists. So I ran a node and checks the ouput of `GET /_nodes/stats` and saw that Discovery stats were located in `nodes.ABC.fs.discovery` meaning that something messed up the object hierarchy.

I think we could detect malformed objects built with `XContentBuilder` and throws an exception. I'll try to have a look at this.
